### PR TITLE
fs: added usage examples/doctests to File

### PR DESCRIPTION
**Pull request by mattgathu**
Motivation
--
Part of [this](https://github.com/rust-lang-nursery/wg-net/issues/54) networking working group issue.

Solution
--
Add doctests to public methods of the `tokio::fs::File` struct.

- file::File
- file::File::from_std
- file::File::create
- file::File::into_std
- file::File::metadata
- file::File::open
- file::File::poll_metadata
- file::File::poll_seek
- file::File::poll_set_len
- file::File::poll_set_permissions
- file::File::poll_sync_all
- file::File::poll_sync_data
- file::File::poll_try_clone
- file::File::seek

### CODE REVIEW

File: tokio-fs/src/file/mod.rs [hunk 1]
```diff
@@ -34,6 +34,46 @@ use std::path::Path;
...
+///
+/// ```no_run
+/// extern crate tokio;
+/// 
+/// use tokio::prelude::*;
+/// 
+/// fn main() {
+///     let task = tokio::fs::File::open("foo.txt")
+///         .and_then(|mut file| {
+///             let mut contents = vec!();
```
Comment (tobz) [inline tokio-fs/src/file/mod.rs]: I think you meant `vec![]`?

File: tokio-fs/src/file/mod.rs [hunk 2]
```diff
@@ -34,6 +34,46 @@ use std::path::Path;
...
 /// [std]: https://doc.rust-lang.org/std/fs/struct.File.html
+///
+/// # Examples
+///
+/// Create a new file and asynchronously write bytes to it:
+///
+/// ```no_run
+/// extern crate tokio;
+/// 
+/// use tokio::prelude::*;
```
Comment (tobz) [inline tokio-fs/src/file/mod.rs]: Do you actually need the prelude here?  Given that the File type is fully qualified?

File: tokio-fs/src/file/mod.rs [hunk 3]
```diff
@@ -73,15 +130,45 @@ impl File {
...
     ///
     /// [`create`]: https://doc.rust-lang.org/std/fs/struct.File.html#method.create
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # extern crate tokio;
+    /// use tokio::prelude::*;
+    /// fn main() {
+    ///     let task = tokio::fs::File::create("foo.txt").and_then(|file| {
```
Comment (tobz) [inline tokio-fs/src/file/mod.rs]: Break the `and_then` to its own line.

File: tokio-fs/src/file/mod.rs [hunk 4]
```diff
@@ -73,15 +130,45 @@ impl File {
...
     ///
     /// [std]: https://doc.rust-lang.org/std/fs/struct.File.html
+    /// [file]: struct.File.html
+    ///
+    /// Examples
+    /// ```no_run
+    /// # extern crate tokio;
+    /// use std::fs::File;
+    /// 
+    /// fn main() -> Result<(), Box<std::error::Error>> {
```
Comment (tobz) [inline tokio-fs/src/file/mod.rs]: I think drop the `Result` (and the corresponding `Ok`) from this.  We aren't going to run it in CI, so we can be a little more terse.

File: tokio-fs/src/file/mod.rs [hunk 5]
```diff
@@ -143,23 +310,96 @@ impl File {
...
+    /// 
+    /// fn main() {
+    /// let task = tokio::fs::File::create("foo.txt")
+    ///     .and_then(|mut file| file.poll_try_clone())
+    ///     .map(|clone| {
+    ///         // do something with the clone
+    ///         # println!("{:?}", clone);
+    ///     }).map_err(|err| eprintln!("IO error {:?}", err));
+    /// 
+    /// tokio::run(task);
```
Comment (tobz) [inline tokio-fs/src/file/mod.rs]: Indent.

File: tokio-fs/src/file/mod.rs [hunk 6]
```diff
@@ -182,6 +422,29 @@ impl File {
...
+    ///
+    /// ```no_run
+    /// # extern crate tokio;
+    /// use tokio::prelude::*;
+    /// 
+    /// fn main() {
+    ///     let task = tokio::fs::File::create("foo.txt")
+    ///         .and_then(|file| file.metadata())
+    ///         .map(|file_metadata| {
+    ///             let (mut file, metadata) = file_metadata;
```
Comment (tobz) [inline tokio-fs/src/file/mod.rs]: I might collapse this line and pull out both `file` and `metadata` right from the closure params.

File: tokio-fs/src/file/mod.rs [hunk 7]
```diff
@@ -52,6 +92,23 @@ impl File {
...
+    ///
+    /// ```no_run
+    /// # extern crate tokio;
+    /// use tokio::prelude::*;
+    /// fn main() {
+    ///     let task = tokio::fs::File::open("foo.txt").and_then(|file| {
+    ///         // do something with the file ...
+    ///         file.metadata().map(|md| println!("{:?}", md))
+    ///     }).map_err(|e| {
+    ///         //handle errors
```
Comment (tobz) [inline tokio-fs/src/file/mod.rs]: Space between `//` and sentence.

File: tokio-fs/src/file/mod.rs [hunk 8]
```diff
@@ -73,15 +130,45 @@ impl File {
...
+    ///
+    /// ```no_run
+    /// # extern crate tokio;
+    /// use tokio::prelude::*;
+    /// fn main() {
+    ///     let task = tokio::fs::File::create("foo.txt").and_then(|file| {
+    ///         // do something with the created file ...
+    ///         file.metadata().map(|md| println!("{:?}", md))
+    ///     }).map_err(|e| {
+    ///         //handle errors
```
Comment (tobz) [inline tokio-fs/src/file/mod.rs]: Space between `//` and sentence.

File: tokio-fs/src/file/mod.rs (hunk 4 above)
Comment (mattgathu) [reply tokio-fs/src/file/mod.rs]: cool. I have dropped the result and now just have a single unwrap.

**Comment by mattgathu** [comment]
@tobz I have updated the pr:

- used specific imports from `prelude` instead of a wildcard.
- refactored IO error `map_err` and `eprintln!`
- refactored out `Result` in one of the examples

[diff](
https://github.com/tokio-rs/tokio/pull/786/commits/31b1bb0a56b387e34ac44ac441dd3f7ceac4c189)

**Comment by tobz** [comment]
Great, I think things are consistent enough now. 👍

**Pull request merged**